File: src/shared/contracts/tools/check-test-matrix.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import csv
import json
import re
import sys
import tomllib
from pathlib import Path


ROOT = Path(__file__).resolve().parents[4]
CONTRACTS_ROOT = ROOT / "src/shared/contracts"
FIXTURES_ROOT = ROOT / "src/shared/fixtures"
MATRIX_PATH = CONTRACTS_ROOT / "test-matrix.toml"
GENERATED_MANIFEST = ROOT / "target/shared-fixtures/manifest.generated.json"
GENERATED_CONSUMPTION_REPORT = ROOT / "target/shared-fixtures/consumption-report.json"
ID_RE = re.compile(r"^[a-z0-9][a-z0-9.-]*[a-z0-9]$")
FORMATS = {"json", "properties", "tsv"}
EVIDENCE_KINDS = {"fixture-file", "semantic-contract"}
CONSUMPTION_SCAN_ROOTS = [
    "src/sdks/rust/tests",
    "src/sdks/swift/Tests",
    "src/sdks/kotlin/oliphaunt/src",
    "src/sdks/js/src",
    "src/sdks/react-native/src",
    "src/bindings/wasix-rust/crates/oliphaunt-wasix/src",
    "tools/release",
]
CODE_SUFFIXES = {
    ".bash",
    ".c",
    ".cjs",
    ".cpp",
    ".gradle",
    ".h",
    ".java",
    ".js",
    ".kt",
    ".kts",
    ".mjs",
    ".mm",
    ".py",
    ".rs",
    ".sh",
    ".swift",
    ".ts",
    ".tsx",
}
IGNORED_DIR_NAMES = {
    ".build",
    ".gradle",
    ".moon",
    ".next",
    "__pycache__",
    "build",
    "DerivedData",
    "dist",
    "lib",
    "node_modules",
    "target",
}
PROJECT_ROOTS = {
    "src/runtimes/liboliphaunt/native": "liboliphaunt-native",
    "src/sdks/rust": "oliphaunt-rust",
    "src/sdks/swift": "oliphaunt-swift",
    "src/sdks/kotlin": "oliphaunt-kotlin",
    "src/sdks/js": "oliphaunt-js",
    "src/sdks/react-native": "oliphaunt-react-native",
    "src/bindings/wasix-rust": "oliphaunt-wasix-rust",
    "tools/policy": "policy-tools",
    "tools/release": "release-tools",
}


def fail(message: str) -> None:
    raise SystemExit(message)


def load_matrix() -> dict:
    try:
        with MATRIX_PATH.open("rb") as handle:
            return tomllib.load(handle)
    except tomllib.TOMLDecodeError as error:
        fail(f"{MATRIX_PATH}: invalid TOML: {error}")


def validate_fixture_entry(entry: dict, seen: set[str]) -> dict:
    fixture_id = require_string(entry, "id")
    if not ID_RE.match(fixture_id):
        fail(f"{MATRIX_PATH}: invalid fixture id {fixture_id!r}")
    if fixture_id in seen:
        fail(f"{MATRIX_PATH}: duplicate fixture id {fixture_id!r}")
    seen.add(fixture_id)

    relative_path = require_string(entry, "path")
    path = Path(relative_path)
    if path.is_absolute() or ".." in path.parts:
        fail(f"{MATRIX_PATH}: fixture {fixture_id} has unsafe path {relative_path!r}")

    fixture_format = require_string(entry, "format")
    if fixture_format not in FORMATS:
        fail(f"{MATRIX_PATH}: fixture {fixture_id} has unsupported format {fixture_format!r}")

    contract = require_string(entry, "contract")
    proof_owner = require_string(entry, "proof_owner")
    ci_tier = require_string(entry, "ci_tier")
    if not re.match(r"^T[0-8]$", ci_tier):
        fail(f"{MATRIX_PATH}: fixture {fixture_id} has invalid ci_tier {ci_tier!r}")
    consumers = entry.get("consumers")
    if not isinstance(consumers, list) or not consumers or not all(isinstance(item, str) and item for item in consumers):
        fail(f"{MATRIX_PATH}: fixture {fixture_id} must declare non-empty string consumers")
    non_consumers = entry.get("non_consumers")
    if not isinstance(non_consumers, list) or not all(isinstance(item, str) and item for item in non_consumers):
        fail(f"{MATRIX_PATH}: fixture {fixture_id} must declare string non_consumers")
    overlap = set(consumers).intersection(non_consumers)
    if overlap:
        fail(f"{MATRIX_PATH}: fixture {fixture_id} declares consumers as non-consumers: {sorted(overlap)}")

    shared = entry.get("shared")
    if not isinstance(shared, bool):
        fail(f"{MATRIX_PATH}: fixture {fixture_id} must declare shared = true/false")
    if shared and len(set(consumers)) < 2:
        fail(f"{MATRIX_PATH}: shared fixture {fixture_id} must have at least two consumers")
    if not shared and not isinstance(entry.get("reason"), str):
        fail(f"{MATRIX_PATH}: product-specific fixture {fixture_id} must explain why it is cataloged")
    evidence = entry.get("evidence", [])
    if not isinstance(evidence, list) or not evidence:
        fail(f"{MATRIX_PATH}: fixture {fixture_id} must declare evidence for every consumer")
    evidence_consumers: list[str] = []
    for item in evidence:
        if not isinstance(item, dict):
            fail(f"{MATRIX_PATH}: fixture {fixture_id} evidence entries must be TOML tables")
        consumer = require_string(item, "consumer")
        if consumer not in consumers:
            fail(f"{MATRIX_PATH}: fixture {fixture_id} has evidence for undeclared consumer {consumer!r}")
        evidence_consumers.append(consumer)
        kind = item.get("kind", "fixture-file")
        if kind not in EVIDENCE_KINDS:
            fail(f"{MATRIX_PATH}: fixture {fixture_id} evidence for {consumer} has unsupported kind {kind!r}")
        evidence_path = require_string(item, "path")
        path = Path(evidence_path)
        if path.is_absolute() or ".." in path.parts:
            fail(f"{MATRIX_PATH}: fixture {fixture_id} evidence for {consumer} has unsafe path {evidence_path!r}")
        markers = item.get("markers")
        if not isinstance(markers, list) or not markers or not all(isinstance(marker, str) and marker for marker in markers):
            fail(f"{MATRIX_PATH}: fixture {fixture_id} evidence for {consumer} must declare non-empty string markers")
    missing_evidence = sorted(set(consumers).difference(evidence_consumers))
    if missing_evidence:
        fail(f"{MATRIX_PATH}: fixture {fixture_id} lacks evidence for consumers: {missing_evidence}")

    return {
        "id": fixture_id,
        "path": relative_path,
        "format": fixture_format,
        "contract": contract,
        "proof_owner": proof_owner,
        "ci_tier": ci_tier,
        "shared": shared,
        "consumers": consumers,
        "non_consumers": non_consumers,
        "evidence": evidence,
    }


def require_string(entry: dict, key: str) -> str:
    value = entry.get(key)
    if not isinstance(value, str) or not value:
        fail(f"{MATRIX_PATH}: fixture entry missing string {key!r}")
    return value


def validate_fixture_file(entry: dict) -> dict:
    relative_path = entry["path"]
    fixture_path = FIXTURES_ROOT / relative_path
    if not fixture_path.is_file():
        fail(f"missing shared fixture {fixture_path}")

    if entry["format"] == "json":
        with fixture_path.open("r", encoding="utf-8") as handle:
            parsed = json.load(handle)
        if not isinstance(parsed, dict):
            fail(f"{fixture_path}: JSON fixture must be an object")
    elif entry["format"] == "properties":
        validate_properties(fixture_path)
    elif entry["format"] == "tsv":
        validate_tsv(fixture_path)

    return {
        "id": entry["id"],
        "path": f"src/shared/fixtures/{relative_path}",
        "format": entry["format"],
        "proofOwner": entry["proof_owner"],
        "ciTier": entry["ci_tier"],
        "consumers": entry["consumers"],
        "nonConsumers": entry["non_consumers"],
        "shared": entry["shared"],
        "evidence": [
            validate_evidence_file(entry, evidence)
            for evidence in entry["evidence"]
        ],
    }


def validate_evidence_file(fixture: dict, evidence: dict) -> dict:
    evidence_path = ROOT / evidence["path"]
    if not evidence_path.is_file():
        fail(f"{MATRIX_PATH}: fixture {fixture['id']} evidence file does not exist: {evidence_path}")
    text = evidence_path.read_text(encoding="utf-8")
    for marker in evidence["markers"]:
        if marker not in text:
            fail(
                f"{MATRIX_PATH}: fixture {fixture['id']} evidence file {evidence['path']} "
                f"for {evidence['consumer']} lacks marker {marker!r}"
            )
    return {
        "consumer": evidence["consumer"],
        "kind": evidence.get("kind", "fixture-file"),
        "path": evidence["path"],
        "markers": evidence["markers"],
    }


def load_project_roots() -> dict[str, str]:
    roots = dict(PROJECT_ROOTS)
    for root, project_id in PROJECT_ROOTS.items():
        moon_file = ROOT / root / "moon.yml"
        if not moon_file.is_file():
            fail(f"{MATRIX_PATH}: fixture matrix project root {root} is missing moon.yml")
        match = re.search(r"(?m)^id:\s*[\"']?([^\"'\s#]+)", moon_file.read_text(encoding="utf-8"))
        if not match:
            fail(f"{MATRIX_PATH}: fixture matrix project root {root} moon.yml has no id")
        actual_project_id = match.group(1)
        if actual_project_id != project_id:
            fail(
                f"{MATRIX_PATH}: fixture matrix project root {root} expected id "
                f"{project_id}, got {actual_project_id}"
            )
    return roots


def project_for_path(path: Path, project_roots: dict[str, str]) -> str | None:
    relative = path.relative_to(ROOT).as_posix()
    best_root = ""
    best_project: str | None = None
    for root, project_id in project_roots.items():
        if relative == root or relative.startswith(f"{root}/"):
            if len(root) > len(best_root):
                best_root = root
                best_project = project_id
    return best_project


def validate_project_ids(entries: list[dict], project_roots: dict[str, str]) -> None:
    known_ids = set(project_roots.values())
    for entry in entries:
        ids = set(entry["consumers"]) | set(entry["non_consumers"])
        ids.update(evidence["consumer"] for evidence in entry["evidence"])
        unknown = sorted(ids.difference(known_ids))
        if unknown:
            fail(f"{MATRIX_PATH}: fixture {entry['id']} references unknown Moon project ids: {unknown}")


def detect_fixture_references(entries: list[dict], project_roots: dict[str, str]) -> list[dict]:
    by_pattern: dict[str, dict] = {}
    for entry in entries:
        relative_path = entry["path"]
        by_pattern[f"src/shared/fixtures/{relative_path}"] = entry
        by_pattern[relative_path] = entry

    detections: list[dict] = []
    seen: set[tuple[str, str, str]] = set()
    for scan_root in CONSUMPTION_SCAN_ROOTS:
        root = ROOT / scan_root
        if not root.exists():
            continue
        for path in root.rglob("*"):
            if not path.is_file() or path.suffix not in CODE_SUFFIXES:
                continue
            relative_parts = path.relative_to(ROOT).parts
            if any(part in IGNORED_DIR_NAMES for part in relative_parts):
                continue
            try:
                text = path.read_text(encoding="utf-8")
            except UnicodeDecodeError:
                continue
            for pattern, entry in by_pattern.items():
                if pattern not in text:
                    continue
                project_id = project_for_path(path, project_roots)
                if project_id is None:
                    fail(f"{MATRIX_PATH}: fixture reference in unmanaged path {path.relative_to(ROOT)}")
                if project_id in entry["non_consumers"] or project_id not in entry["consumers"]:
                    fail(
                        f"{MATRIX_PATH}: {project_id} references fixture {entry['id']} "
                        f"from {path.relative_to(ROOT)}, but allowed consumers are {entry['consumers']}"
                    )
                detection_key = (entry["id"], project_id, path.relative_to(ROOT).as_posix())
                if detection_key in seen:
                    continue
                seen.add(detection_key)
                detections.append(
                    {
                        "fixtureId": entry["id"],
                        "project": project_id,
                        "path": path.relative_to(ROOT).as_posix(),
                        "matched": pattern,
                    }
                )
    return detections


def write_consumption_report(entries: list[dict], detections: list[dict]) -> None:
    detections_by_fixture: dict[str, list[dict]] = {entry["id"]: [] for entry in entries}
    for detection in detections:
        detections_by_fixture.setdefault(detection["fixtureId"], []).append(detection)

    report = {
        "schemaVersion": 1,
        "fixtures": [
            {
                "id": entry["id"],
                "path": f"src/shared/fixtures/{entry['path']}",
                "consumers": entry["consumers"],
                "evidence": [
                    {
                        "consumer": evidence["consumer"],
                        "kind": evidence.get("kind", "fixture-file"),
                        "path": evidence["path"],
                    }
                    for evidence in entry["evidence"]
                ],
                "detectedReferences": detections_by_fixture.get(entry["id"], []),
            }
            for entry in entries
        ],
    }
    GENERATED_CONSUMPTION_REPORT.parent.mkdir(parents=True, exist_ok=True)
    GENERATED_CONSUMPTION_REPORT.write_text(
        json.dumps(report, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )


def validate_properties(path: Path) -> None:
    lines = path.read_text(encoding="utf-8").splitlines()
    entries = [
        line
        for line in lines
        if line.strip() and not line.lstrip().startswith("#")
    ]
    if not entries:
        fail(f"{path}: properties fixture is empty")
    for line in entries:
        if "=" not in line:
            fail(f"{path}: properties line lacks '=': {line!r}")


def validate_tsv(path: Path) -> None:
    with path.open("r", encoding="utf-8", newline="") as handle:
        rows = list(csv.reader(handle, delimiter="\t"))
    if len(rows) < 2:
        fail(f"{path}: TSV fixture must contain a header and at least one data row")
    width = len(rows[0])
    if width == 0:
        fail(f"{path}: TSV fixture header is empty")
    for index, row in enumerate(rows[1:], start=2):
        if len(row) != width:
            fail(f"{path}: row {index} has {len(row)} cells, expected {width}")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--fixtures",
        action="store_true",
        help="also validate fixture files and emit the generated manifest",
    )
    args = parser.parse_args()

    matrix = load_matrix()
    if matrix.get("schema_version") != 1:
        fail(f"{MATRIX_PATH}: schema_version must be 1")
    raw_fixtures = matrix.get("fixtures")
    if not isinstance(raw_fixtures, list) or not raw_fixtures:
        fail(f"{MATRIX_PATH}: must declare at least one [[fixtures]] entry")

    seen: set[str] = set()
    entries = [validate_fixture_entry(entry, seen) for entry in raw_fixtures]

    if args.fixtures:
        project_roots = load_project_roots()
        validate_project_ids(entries, project_roots)
        detections = detect_fixture_references(entries, project_roots)
        generated = {
            "schemaVersion": 1,
            "fixtures": [validate_fixture_file(entry) for entry in entries],
        }
        GENERATED_MANIFEST.parent.mkdir(parents=True, exist_ok=True)
        GENERATED_MANIFEST.write_text(json.dumps(generated, indent=2, sort_keys=True) + "\n", encoding="utf-8")
        write_consumption_report(entries, detections)

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/shared/contracts/tools/check-test-matrix.mjs
#!/usr/bin/env bun
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..', '..', '..');
const CONTRACTS_ROOT = path.join(ROOT, 'src/shared/contracts');
const FIXTURES_ROOT = path.join(ROOT, 'src/shared/fixtures');
const MATRIX_PATH = path.join(CONTRACTS_ROOT, 'test-matrix.toml');
const GENERATED_MANIFEST = path.join(ROOT, 'target/shared-fixtures/manifest.generated.json');
const GENERATED_CONSUMPTION_REPORT = path.join(ROOT, 'target/shared-fixtures/consumption-report.json');
const ID_RE = /^[a-z0-9][a-z0-9.-]*[a-z0-9]$/u;
const FORMATS = new Set(['json', 'properties', 'tsv']);
const EVIDENCE_KINDS = new Set(['fixture-file', 'semantic-contract']);
const CONSUMPTION_SCAN_ROOTS = [
  'src/sdks/rust/tests',
  'src/sdks/swift/Tests',
  'src/sdks/kotlin/oliphaunt/src',
  'src/sdks/js/src',
  'src/sdks/react-native/src',
  'src/bindings/wasix-rust/crates/oliphaunt-wasix/src',
  'tools/release',
];
const CODE_SUFFIXES = new Set([
  '.bash',
  '.c',
  '.cjs',
  '.cpp',
  '.gradle',
  '.h',
  '.java',
  '.js',
  '.kt',
  '.kts',
  '.mjs',
  '.mm',
  '.py',
  '.rs',
  '.sh',
  '.swift',
  '.ts',
  '.tsx',
]);
const IGNORED_DIR_NAMES = new Set([
  '.build',
  '.gradle',
  '.moon',
  '.next',
  '__pycache__',
  'build',
  'DerivedData',
  'dist',
  'lib',
  'node_modules',
  'target',
]);
const PROJECT_ROOTS = {
  'src/runtimes/liboliphaunt/native': 'liboliphaunt-native',
  'src/sdks/rust': 'oliphaunt-rust',
  'src/sdks/swift': 'oliphaunt-swift',
  'src/sdks/kotlin': 'oliphaunt-kotlin',
  'src/sdks/js': 'oliphaunt-js',
  'src/sdks/react-native': 'oliphaunt-react-native',
  'src/bindings/wasix-rust': 'oliphaunt-wasix-rust',
  'tools/policy': 'policy-tools',
  'tools/release': 'release-tools',
};

function fail(message) {
  console.error(message);
  process.exit(1);
}

function posixRelative(file) {
  return path.relative(ROOT, file).split(path.sep).join('/');
}

function isPlainObject(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function stableValue(value) {
  if (Array.isArray(value)) {
    return value.map(stableValue);
  }
  if (!isPlainObject(value)) {
    return value;
  }
  const sorted = {};
  for (const key of Object.keys(value).sort()) {
    sorted[key] = stableValue(value[key]);
  }
  return sorted;
}

function stableJson(value) {
  return `${JSON.stringify(stableValue(value), null, 2)}\n`;
}

function readText(file) {
  return fs.readFileSync(file, 'utf8');
}

function requireString(entry, key) {
  const value = entry?.[key];
  if (typeof value !== 'string' || value.length === 0) {
    fail(`${MATRIX_PATH}: fixture entry missing string ${JSON.stringify(key)}`);
  }
  return value;
}

function isSafeRelative(relativePath) {
  const parts = relativePath.split(/[\\/]/u);
  return !path.isAbsolute(relativePath) && !parts.includes('..');
}

function loadMatrix() {
  try {
    return Bun.TOML.parse(readText(MATRIX_PATH));
  } catch (error) {
    fail(`${MATRIX_PATH}: invalid TOML: ${error.message}`);
  }
}

function validateFixtureEntry(entry, seen) {
  const fixtureId = requireString(entry, 'id');
  if (!ID_RE.test(fixtureId)) {
    fail(`${MATRIX_PATH}: invalid fixture id ${JSON.stringify(fixtureId)}`);
  }
  if (seen.has(fixtureId)) {
    fail(`${MATRIX_PATH}: duplicate fixture id ${JSON.stringify(fixtureId)}`);
  }
  seen.add(fixtureId);

  const relativePath = requireString(entry, 'path');
  if (!isSafeRelative(relativePath)) {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} has unsafe path ${JSON.stringify(relativePath)}`);
  }

  const fixtureFormat = requireString(entry, 'format');
  if (!FORMATS.has(fixtureFormat)) {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} has unsupported format ${JSON.stringify(fixtureFormat)}`);
  }

  const contract = requireString(entry, 'contract');
  const proofOwner = requireString(entry, 'proof_owner');
  const ciTier = requireString(entry, 'ci_tier');
  if (!/^T[0-8]$/u.test(ciTier)) {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} has invalid ci_tier ${JSON.stringify(ciTier)}`);
  }

  const consumers = entry.consumers;
  if (!Array.isArray(consumers) || consumers.length === 0 || !consumers.every((item) => typeof item === 'string' && item.length > 0)) {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} must declare non-empty string consumers`);
  }
  const nonConsumers = entry.non_consumers;
  if (!Array.isArray(nonConsumers) || !nonConsumers.every((item) => typeof item === 'string' && item.length > 0)) {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} must declare string non_consumers`);
  }
  const overlap = consumers.filter((consumer) => nonConsumers.includes(consumer)).sort();
  if (overlap.length > 0) {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} declares consumers as non-consumers: ${JSON.stringify(overlap)}`);
  }

  const shared = entry.shared;
  if (typeof shared !== 'boolean') {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} must declare shared = true/false`);
  }
  if (shared && new Set(consumers).size < 2) {
    fail(`${MATRIX_PATH}: shared fixture ${fixtureId} must have at least two consumers`);
  }
  if (!shared && typeof entry.reason !== 'string') {
    fail(`${MATRIX_PATH}: product-specific fixture ${fixtureId} must explain why it is cataloged`);
  }

  const evidence = entry.evidence ?? [];
  if (!Array.isArray(evidence) || evidence.length === 0) {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} must declare evidence for every consumer`);
  }
  const evidenceConsumers = [];
  for (const item of evidence) {
    if (!isPlainObject(item)) {
      fail(`${MATRIX_PATH}: fixture ${fixtureId} evidence entries must be TOML tables`);
    }
    const consumer = requireString(item, 'consumer');
    if (!consumers.includes(consumer)) {
      fail(`${MATRIX_PATH}: fixture ${fixtureId} has evidence for undeclared consumer ${JSON.stringify(consumer)}`);
    }
    evidenceConsumers.push(consumer);
    const kind = item.kind ?? 'fixture-file';
    if (!EVIDENCE_KINDS.has(kind)) {
      fail(`${MATRIX_PATH}: fixture ${fixtureId} evidence for ${consumer} has unsupported kind ${JSON.stringify(kind)}`);
    }
    const evidencePath = requireString(item, 'path');
    if (!isSafeRelative(evidencePath)) {
      fail(`${MATRIX_PATH}: fixture ${fixtureId} evidence for ${consumer} has unsafe path ${JSON.stringify(evidencePath)}`);
    }
    const markers = item.markers;
    if (!Array.isArray(markers) || markers.length === 0 || !markers.every((marker) => typeof marker === 'string' && marker.length > 0)) {
      fail(`${MATRIX_PATH}: fixture ${fixtureId} evidence for ${consumer} must declare non-empty string markers`);
    }
  }
  const missingEvidence = consumers.filter((consumer) => !evidenceConsumers.includes(consumer)).sort();
  if (missingEvidence.length > 0) {
    fail(`${MATRIX_PATH}: fixture ${fixtureId} lacks evidence for consumers: ${JSON.stringify(missingEvidence)}`);
  }

  return {
    id: fixtureId,
    path: relativePath,
    format: fixtureFormat,
    contract,
    proof_owner: proofOwner,
    ci_tier: ciTier,
    shared,
    consumers,
    non_consumers: nonConsumers,
    evidence,
  };
}

function validateProperties(file) {
  const entries = readText(file)
    .split(/\r?\n/u)
    .filter((line) => line.trim().length > 0 && !line.trimStart().startsWith('#'));
  if (entries.length === 0) {
    fail(`${file}: properties fixture is empty`);
  }
  for (const line of entries) {
    if (!line.includes('=')) {
      fail(`${file}: properties line lacks '=': ${JSON.stringify(line)}`);
    }
  }
}

function parseTsvLine(line) {
  const cells = [];
  let cell = '';
  let quoted = false;
  for (let index = 0; index < line.length; index += 1) {
    const char = line[index];
    if (char === '"') {
      if (quoted && line[index + 1] === '"') {
        cell += '"';
        index += 1;
      } else {
        quoted = !quoted;
      }
      continue;
    }
    if (char === '\t' && !quoted) {
      cells.push(cell);
      cell = '';
      continue;
    }
    cell += char;
  }
  cells.push(cell);
  return cells;
}

function validateTsv(file) {
  const rows = readText(file)
    .replace(/\r\n/gu, '\n')
    .replace(/\r/gu, '\n')
    .split('\n')
    .filter((line, index, lines) => index < lines.length - 1 || line.length > 0)
    .map(parseTsvLine);
  if (rows.length < 2) {
    fail(`${file}: TSV fixture must contain a header and at least one data row`);
  }
  const width = rows[0].length;
  if (width === 0) {
    fail(`${file}: TSV fixture header is empty`);
  }
  rows.slice(1).forEach((row, index) => {
    if (row.length !== width) {
      fail(`${file}: row ${index + 2} has ${row.length} cells, expected ${width}`);
    }
  });
}

function validateEvidenceFile(fixture, evidence) {
  const evidencePath = path.join(ROOT, evidence.path);
  if (!fs.existsSync(evidencePath) || !fs.statSync(evidencePath).isFile()) {
    fail(`${MATRIX_PATH}: fixture ${fixture.id} evidence file does not exist: ${evidencePath}`);
  }
  const text = readText(evidencePath);
  for (const marker of evidence.markers) {
    if (!text.includes(marker)) {
      fail(
        `${MATRIX_PATH}: fixture ${fixture.id} evidence file ${evidence.path} ` +
          `for ${evidence.consumer} lacks marker ${JSON.stringify(marker)}`,
      );
    }
  }
  return {
    consumer: evidence.consumer,
    kind: evidence.kind ?? 'fixture-file',
    path: evidence.path,
    markers: evidence.markers,
  };
}

function validateFixtureFile(entry) {
  const fixturePath = path.join(FIXTURES_ROOT, entry.path);
  if (!fs.existsSync(fixturePath) || !fs.statSync(fixturePath).isFile()) {
    fail(`missing shared fixture ${fixturePath}`);
  }

  if (entry.format === 'json') {
    const parsed = JSON.parse(readText(fixturePath));
    if (!isPlainObject(parsed)) {
      fail(`${fixturePath}: JSON fixture must be an object`);
    }
  } else if (entry.format === 'properties') {
    validateProperties(fixturePath);
  } else if (entry.format === 'tsv') {
    validateTsv(fixturePath);
  }

  return {
    id: entry.id,
    path: `src/shared/fixtures/${entry.path}`,
    format: entry.format,
    proofOwner: entry.proof_owner,
    ciTier: entry.ci_tier,
    consumers: entry.consumers,
    nonConsumers: entry.non_consumers,
    shared: entry.shared,
    evidence: entry.evidence.map((evidence) => validateEvidenceFile(entry, evidence)),
  };
}

function loadProjectRoots() {
  const roots = { ...PROJECT_ROOTS };
  for (const [root, projectId] of Object.entries(PROJECT_ROOTS)) {
    const moonFile = path.join(ROOT, root, 'moon.yml');
    if (!fs.existsSync(moonFile) || !fs.statSync(moonFile).isFile()) {
      fail(`${MATRIX_PATH}: fixture matrix project root ${root} is missing moon.yml`);
    }
    const match = readText(moonFile).match(/^id:\s*["']?([^"'\s#]+)/mu);
    if (match === null) {
      fail(`${MATRIX_PATH}: fixture matrix project root ${root} moon.yml has no id`);
    }
    const actualProjectId = match[1];
    if (actualProjectId !== projectId) {
      fail(`${MATRIX_PATH}: fixture matrix project root ${root} expected id ${projectId}, got ${actualProjectId}`);
    }
  }
  return roots;
}

function projectForPath(file, projectRoots) {
  const relative = posixRelative(file);
  let bestRoot = '';
  let bestProject = null;
  for (const [root, projectId] of Object.entries(projectRoots)) {
    if (relative === root || relative.startsWith(`${root}/`)) {
      if (root.length > bestRoot.length) {
        bestRoot = root;
        bestProject = projectId;
      }
    }
  }
  return bestProject;
}

function validateProjectIds(entries, projectRoots) {
  const knownIds = new Set(Object.values(projectRoots));
  for (const entry of entries) {
    const ids = new Set([
      ...entry.consumers,
      ...entry.non_consumers,
      ...entry.evidence.map((evidence) => evidence.consumer),
    ]);
    const unknown = [...ids].filter((id) => !knownIds.has(id)).sort();
    if (unknown.length > 0) {
      fail(`${MATRIX_PATH}: fixture ${entry.id} references unknown Moon project ids: ${JSON.stringify(unknown)}`);
    }
  }
}

function* walkFiles(root) {
  if (!fs.existsSync(root)) {
    return;
  }
  const entries = fs.readdirSync(root, { withFileTypes: true }).sort((left, right) => left.name.localeCompare(right.name));
  for (const entry of entries) {
    const file = path.join(root, entry.name);
    if (entry.isDirectory()) {
      if (!IGNORED_DIR_NAMES.has(entry.name)) {
        yield* walkFiles(file);
      }
      continue;
    }
    if (entry.isFile()) {
      yield file;
    }
  }
}

function detectFixtureReferences(entries, projectRoots) {
  const byPattern = new Map();
  for (const entry of entries) {
    byPattern.set(`src/shared/fixtures/${entry.path}`, entry);
    byPattern.set(entry.path, entry);
  }

  const detections = [];
  const seen = new Set();
  for (const scanRoot of CONSUMPTION_SCAN_ROOTS) {
    for (const file of walkFiles(path.join(ROOT, scanRoot))) {
      if (!CODE_SUFFIXES.has(path.extname(file))) {
        continue;
      }
      const relativeParts = posixRelative(file).split('/');
      if (relativeParts.some((part) => IGNORED_DIR_NAMES.has(part))) {
        continue;
      }
      let text;
      try {
        text = readText(file);
      } catch (error) {
        if (error instanceof TypeError) {
          continue;
        }
        throw error;
      }
      for (const [pattern, entry] of byPattern.entries()) {
        if (!text.includes(pattern)) {
          continue;
        }
        const projectId = projectForPath(file, projectRoots);
        if (projectId === null) {
          fail(`${MATRIX_PATH}: fixture reference in unmanaged path ${posixRelative(file)}`);
        }
        if (entry.non_consumers.includes(projectId) || !entry.consumers.includes(projectId)) {
          fail(
            `${MATRIX_PATH}: ${projectId} references fixture ${entry.id} from ${posixRelative(file)}, ` +
              `but allowed consumers are ${JSON.stringify(entry.consumers)}`,
          );
        }
        const detectionKey = `${entry.id}\0${projectId}\0${posixRelative(file)}`;
        if (seen.has(detectionKey)) {
          continue;
        }
        seen.add(detectionKey);
        detections.push({
          fixtureId: entry.id,
          project: projectId,
          path: posixRelative(file),
          matched: pattern,
        });
      }
    }
  }
  return detections;
}

function writeConsumptionReport(entries, detections) {
  const detectionsByFixture = new Map(entries.map((entry) => [entry.id, []]));
  for (const detection of detections) {
    if (!detectionsByFixture.has(detection.fixtureId)) {
      detectionsByFixture.set(detection.fixtureId, []);
    }
    detectionsByFixture.get(detection.fixtureId).push(detection);
  }

  const report = {
    schemaVersion: 1,
    fixtures: entries.map((entry) => ({
      id: entry.id,
      path: `src/shared/fixtures/${entry.path}`,
      consumers: entry.consumers,
      evidence: entry.evidence.map((evidence) => ({
        consumer: evidence.consumer,
        kind: evidence.kind ?? 'fixture-file',
        path: evidence.path,
      })),
      detectedReferences: detectionsByFixture.get(entry.id) ?? [],
    })),
  };
  fs.mkdirSync(path.dirname(GENERATED_CONSUMPTION_REPORT), { recursive: true });
  fs.writeFileSync(GENERATED_CONSUMPTION_REPORT, stableJson(report), 'utf8');
}

function parseArgs(argv) {
  let fixtures = false;
  for (const arg of argv) {
    if (arg === '--fixtures') {
      fixtures = true;
    } else {
      fail(`unknown argument: ${arg}`);
    }
  }
  return { fixtures };
}

const args = parseArgs(Bun.argv.slice(2));
const matrix = loadMatrix();
if (matrix.schema_version !== 1) {
  fail(`${MATRIX_PATH}: schema_version must be 1`);
}
const rawFixtures = matrix.fixtures;
if (!Array.isArray(rawFixtures) || rawFixtures.length === 0) {
  fail(`${MATRIX_PATH}: must declare at least one [[fixtures]] entry`);
}

const seen = new Set();
const entries = rawFixtures.map((entry) => validateFixtureEntry(entry, seen));

if (args.fixtures) {
  const projectRoots = loadProjectRoots();
  validateProjectIds(entries, projectRoots);
  const detections = detectFixtureReferences(entries, projectRoots);
  const generated = {
    schemaVersion: 1,
    fixtures: entries.map(validateFixtureFile),
  };
  fs.mkdirSync(path.dirname(GENERATED_MANIFEST), { recursive: true });
  fs.writeFileSync(GENERATED_MANIFEST, stableJson(generated), 'utf8');
  writeConsumptionReport(entries, detections);
}
